src/library/scala/collection/mutable/Growable.scala

```diff
@@ -56,10 +56,15 @@ trait Growable[-A] extends Clearable {
    *  @param xs   the IterableOnce producing the elements to $add.
    *  @return  the $coll itself.
    */
+  @nowarn("msg=will most likely never compare equal")
   def addAll(xs: IterableOnce[A]): this.type = {
-    val it = xs.iterator
-    while (it.hasNext) {
-      addOne(it.next())
+    xs match {
+      case xs: AnyRef if xs eq this => addAll(Buffer.from(xs)) // avoid mutating under our own iterator
```

Comment:
nitpick: elsewhere we tend to write `if (xs.asInstanceOf[AnyRef] eq this)` which is the same from a code gen perspective. The no-op `asInstanceOf[AnyRef]` isn't pretty but it would be at least consistently so.

Reply:
done